File: project/backend-services/sharespot-device-records-master-backend/infrastructure/persistence/postgres/src/main/java/sharespot/services/devicerecordsbackend/infrastructure/persistence/postgres/mapper/RecordMapper.java
package sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.mapper;

import sharespot.services.devicerecordsbackend.domain.model.device.DeviceId;
import sharespot.services.devicerecordsbackend.domain.model.records.*;
import sharespot.services.devicerecordsbackend.domain.model.subDevices.DeviceRef;
import sharespot.services.devicerecordsbackend.domain.model.subDevices.SubDevice;
import sharespot.services.devicerecordsbackend.domain.model.subDevices.SubDevices;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordEntryPostgres;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordEntryTypePostgres;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceRecordsPostgres;
import sharespot.services.devicerecordsbackend.infrastructure.persistence.postgres.model.DeviceSubSensorPostgres;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

public class RecordMapper {
    public static DeviceRecordsPostgres domainToPostgres(DeviceRecords records) {
        var postgres = new DeviceRecordsPostgres();
        postgres.deviceId = records.device().id().value().toString();
        postgres.name = records.device().name().value();
        postgres.entries = records.records().entries().stream().map(e -> {
            var entry = new DeviceRecordEntryPostgres();
            if (e instanceof BasicRecordEntry) {
                entry.type = DeviceRecordEntryTypePostgres.basic();
            } else {
                entry.type = DeviceRecordEntryTypePostgres.sensorData();
            }
            entry.content = e.getContent();
            entry.label = e.getLabel();
            entry.records = postgres;
            return entry;
        }).collect(Collectors.toSet());

        postgres.subSensors = records.subDevices().entries().stream().map(sub -> {
            var entry = new DeviceSubSensorPostgres();
            entry.subDeviceRef = sub.ref().value();
            entry.subDeviceId = sub.id().value().toString();
            entry.controller = postgres;
            return entry;
        }).collect(Collectors.toSet());

        return postgres;
    }

    public static DeviceRecords postgresToDomain(DeviceRecordsPostgres records) {
        List<RecordEntry> collect = records.entries.stream().map(e -> {
            if (e.type.equals(DeviceRecordEntryTypePostgres.sensorData())) {
                var label = SensorDataRecordLabel.give(e.label);
                return new SensorDataRecordEntry(label, e.content);
            } else {
                return new BasicRecordEntry(e.label, e.content);
            }
        }).collect(Collectors.toList());

        var deviceId = new DeviceId(UUID.fromString(records.deviceId));
        var deviceName = new DeviceName(records.name);

        var subDevices = records.subSensors.stream()
                .map(sub -> new SubDevice(new DeviceId(UUID.fromString(sub.subDeviceId)), new DeviceRef(sub.subDeviceRef)))
                .collect(Collectors.toSet());

        return new DeviceRecords(new Device(deviceId, deviceName), new Records(collect), new SubDevices(subDevices));
    }
}

File: project/frontend-services/libs/device-records/mapper/src/lib/mappers/RecordMapper.ts
import {DeviceRecordDTO, RecordTypeDTO, SensorDataRecordLabelDTO,} from '@frontend-services/device-records/dto';
import {
  Device,
  DeviceRecord,
  RecordEntry,
  RecordEntryType,
  SensorDataRecordLabel,
  SubDevice
} from '@frontend-services/device-records/model';

export class RecordMapper {
  static dtoToModel(dto: DeviceRecordDTO): DeviceRecord {
    const entries = dto.entries.map((e) => {
      if (e.type == RecordTypeDTO.BASIC) {
        return new RecordEntry(e.label, e.content, RecordEntryType.BASIC);
      } else {
        if (
          (e.label as SensorDataRecordLabelDTO) ===
          SensorDataRecordLabelDTO.GPS_LONGITUDE
        ) {
          return new RecordEntry(
            SensorDataRecordLabel.GPS_LONGITUDE,
            e.content,
            RecordEntryType.SENSOR_DATA
          );
        } else {
          return new RecordEntry(
            SensorDataRecordLabel.GPS_LATITUDE,
            e.content,
            RecordEntryType.SENSOR_DATA
          );
        }
      }
    });
    const subDevices = dto.subDevices.map(deviceRef => new SubDevice(deviceRef.ref, deviceRef.id));

    const device = new Device(dto.device.id, dto.device.name);
    return new DeviceRecord(device, entries, subDevices);
  }

  static modelToDto(model: DeviceRecord): DeviceRecordDTO {
    const entries = model.entries.map((e) => {
      if (e.type == RecordEntryType.BASIC) {
        return {
          label: e.label,
          content: e.content,
          type: RecordTypeDTO.BASIC,
        };
      } else {
        if (
          (e.label as SensorDataRecordLabel) ===
          SensorDataRecordLabel.GPS_LONGITUDE
        ) {
          return {
            label: SensorDataRecordLabelDTO.GPS_LONGITUDE,
            content: e.content,
            type: RecordTypeDTO.SENSOR_DATA,
          };
        } else {
          return {
            label: SensorDataRecordLabelDTO.GPS_LATITUDE,
            content: e.content,
            type: RecordTypeDTO.SENSOR_DATA,
          };
        }
      }
    });
    const subDevices = model.subDevices.map(sub => {
      return {
        id: sub.id,
        ref: sub.reference
      }
    })
    return {
      device: {id: model.device.id, name: model.device.name},
      entries: entries,
      subDevices: subDevices
    };
  }
}
